File: domain/src/main/java/js/task/domain/usecase/GetDataUseCase.java
package js.task.domain.usecase;

import androidx.lifecycle.MutableLiveData;

import java.util.List;

import io.reactivex.rxjava3.disposables.CompositeDisposable;
import io.reactivex.rxjava3.schedulers.Schedulers;
import js.task.domain.interfaces.IDataProvider;
import js.task.domain.usecase.interfaces.IGetDataUseCase;
import js.task.domain.usecase.model.DomainModel;


public class GetDataUseCase
        implements IGetDataUseCase
{
    private final IDataProvider iDataProvider;
    private final CompositeDisposable compositeDisposable;

    public GetDataUseCase(
            IDataProvider iDataProvider,
            CompositeDisposable compositeDisposable
    )
    {
        this.iDataProvider = iDataProvider;
        this.compositeDisposable = compositeDisposable;
    }

    @Override
    public void invoke(MutableLiveData<List<DomainModel>> dataOut)
    {
        compositeDisposable.add(iDataProvider.getData()
                .observeOn(Schedulers.io())
                .subscribeOn(Schedulers.io())
                .subscribe(data ->
                {
                    if (data.isEmpty())
                    {
                        iDataProvider.download();
                    }
                    else
                    {
                        dataOut.postValue(data);
                    }
                }));
    }
}

File: domain/src/main/java/js/task/domain/usecase/GetDataUseCase.kt
package js.task.domain.usecase

import js.task.domain.interfaces.IDataProvider
import js.task.domain.usecase.interfaces.IGetDataUseCase
import js.task.domain.usecase.model.DomainModel
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.flow.MutableSharedFlow
import kotlinx.coroutines.launch
import javax.inject.Inject


class GetDataUseCase @Inject constructor(
    private val coroutineScope : CoroutineScope = CoroutineScope(Dispatchers.IO),
    private val dataProvider : IDataProvider
) : IGetDataUseCase
{
    override fun invoke(data : MutableSharedFlow<List<DomainModel>>)
    {
        coroutineScope.launch {
            dataProvider.getData()
                .collect {
                    if (it.isEmpty())
                    {
                        dataProvider.download()
                    }
                    else
                    {
                        data.emit(it)
                    }
                }
        }
    }
}